### PR TITLE
Stop using getTransform

### LINKED ISSUE #861: Firefox ESR doesn't support `context.getTransform`
### Stack trace

```
m.getTransform is not a function:

getTransform (renderer/renderScene.ts:47:42)
componentDidUpdate (index.tsx:2301:37)
p.focus&&p.focus();for (../node_modules/react-dom/cjs/react-dom.production.min.js:251:462)
unstable_runWithPriority (../node_modules/scheduler/cjs/scheduler.production.min.js:18:436)
dg (../node_modules/react-dom/cjs/react-dom.production.min.js:120:324)
Xj (../node_modules/react-dom/cjs/react-dom.production.min.js:244:436)
Lj (../node_modules/react-dom/cjs/react-dom.production.min.js:223:419)
ig (../node_modules/react-dom/cjs/react-dom.production.min.js:121:114)
unstable_runWithPriority (../node_modules/scheduler/cjs/scheduler.production.min.js:18:436)
dg (../node_modules/react-dom/cjs/react-dom.production.min.js:120:324)
ig (../node_modules/react-dom/cjs/react-dom.production.min.js:121:60)
dg (../node_modules/react-dom/cjs/react-dom.production.min.js:120:495)
Rg (../node_modules/react-dom/cjs/react-dom.production.min.js:214:155)
isMounted (../node_modules/react-dom/cjs/react-dom.production.min.js:133:236)
setState (../node_modules/react/cjs/react.production.min.js:13:230)
syncActionResult (index.tsx:704:11)
json=([0-9]+), (index.tsx:774:13)
tryCatch (../node_modules/regenerator-runtime/runtime.js:45:39)
context.dispatchException (../node_modules/regenerator-runtime/runtime.js:271:21)
function (../node_modules/regenerator-runtime/runtime.js:97:20)
asyncGeneratorStep (../node_modules/babel-preset-react-app/node_modules/@babel/runtime/helpers/esm/asyncToGenerator.js:3:24)
_next (../node_modules/babel-preset-react-app/node_modules/@babel/runtime/helpers/esm/asyncToGenerator.js:25:8)
```

### localStorage

```
{"i18nextLng":"en"}
```

### PR DESCRIPTION
**Pull request by vjeux**
Fixes #861

The original motivation behind this is to make it work with Firefox. But it also helped make the code more intentional.

Test Plan:
- Create one square, select it, zoom in repeatedly, make sure that it zooms centered in the screen and everything looks good
- Scroll at various zoom levels, things look good
- Export a small scene at 1x and 3x, make sure the background is properly set and look good

### CODE REVIEW

File: src/renderer/renderScene.ts [hunk 1]
```diff
@@ -51,46 +51,11 @@ export function renderScene(
...
-    );
-  }
-  function resetZoom(context: CanvasRenderingContext2D): void {
-    context.restore();
-  }
+  const normalizedCanvasWidth = canvas.width / scale;
+  const normalizedCanvasHeight = canvas.height / scale;
 
   // Paint background
-  context.save();
```
Comment (vjeux) [inline src/renderer/renderScene.ts]: I removed all the instances of context.save() in favor of explicit set and restore of all the properties

File: src/renderer/renderScene.ts [hunk 2]
```diff
@@ -116,31 +89,26 @@ export function renderScene(
     ),
   );
 
-  applyZoom(context);
```
Comment (vjeux) [inline src/renderer/renderScene.ts]: This was needlessly applying and resetting around each section. We only need to apply and restore once

File: src/renderer/renderScene.ts [hunk 3]
```diff
@@ -166,11 +134,10 @@ export function renderScene(
       context.lineWidth = lineWidth;
       context.setLineDash(initialLineDash);
     });
-    resetZoom(context);
+    context.translate(-sceneState.scrollX, -sceneState.scrollY);
```
Comment (vjeux) [inline src/renderer/renderScene.ts]: The initial translate got reset via context.reset, which is not the most expected behavior

File: src/renderer/renderScene.ts [hunk 4]
```diff
@@ -183,10 +150,14 @@ export function renderScene(
           context.strokeRect(handler[0], handler[1], handler[2], handler[3]);
           context.lineWidth = lineWidth;
         });
-      resetZoom(context);
+      context.translate(-sceneState.scrollX, -sceneState.scrollY);
```
Comment (vjeux) [inline src/renderer/renderScene.ts]: We probably don't need to do the scroll translate twice, but this diff is big enough that I didn't want to touch it

File: src/scene/zoom.ts [hunk 5]
```diff
@@ -16,19 +16,6 @@ export function getZoomOrigin(canvas: HTMLCanvasElement | null) {
   };
 }
 
-export function getZoomTranslation(canvas: HTMLCanvasElement, zoom: number) {
-  const diffMiddleOfTheCanvas = {
-    x: (canvas.width / 2) * (zoom - 1),
-    y: (canvas.height / 2) * (zoom - 1),
-  };
-
-  // Due to JavaScript float precision, we fix to fix decimals count to have symmetric zoom
```
Comment (vjeux) [inline src/scene/zoom.ts]: I have no idea why we did all this hackery. I don't think it's needed

**Comment by Fausto95** [comment]
bugg

I'm running into this bug

File: src/renderer/renderScene.ts [hunk 6]
```diff
@@ -51,46 +51,11 @@ export function renderScene(
 
   const context = canvas.getContext("2d")!;
 
-  // Get initial scale transform as reference for later usage
-  const initialContextTransform = context.getTransform();
```
Comment (vjeux) [inline src/renderer/renderScene.ts]: This was mostly used for getting the pixel ratio of the device. The canvas is scaled outside of this function

File: src/renderer/renderScene.ts [hunk 7]
```diff
@@ -51,46 +51,11 @@ export function renderScene(
...
-    context.setTransform(
-      getContextTransformScaleX(context.getTransform()),
-      0,
-      0,
-      getContextTransformScaleY(context.getTransform()),
-      getContextTransformTranslateX(initialContextTransform) -
-        zoomTranslation.x,
-      getContextTransformTranslateY(initialContextTransform) -
-        zoomTranslation.y,
-    );
```
Comment (vjeux) [inline src/renderer/renderScene.ts]: This is a very convoluted way to do context.translate() and context.scale()

**Comment by vjeux** [comment]
Fausto: I’m going out for a while. If you could investigate that’d be awesome, otherwise i’ll check when back.

It’s likely a transform we don’t reset properly now that resetting is explicit

**Comment by vjeux** [comment]
Good catch, fixed the problem!

**Comment by enzoferey** [comment]
Sorry for not looking at it earlier, but looks good 👍

**Comment by vjeux** [comment]
No worries :)

**Pull request merged**